### PR TITLE
Chore: Hacky first pass on generating composable kinds within grafana/schema

### PR DESCRIPTION
**Pull request by sdboyer**
**What is this feature?**

This adds generation of plugin composable kinds in a central location, within grafana-schema. It is expected, eventually, that all of this move out into grok.

**Why do we need this feature?**

@dprokop has outlined a path by which we use the ~same generated types patterns within scenes and the new frontend experience as we do for as-code. i am _breathtakingly_ excited about where this goes :)

**Who is this feature for?**

scenes folks, presenting a new set of TS interfaces

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer**:

This is EXTREMELY rough. Basically only up to facilitate comment on what's now being generated under `packages/grafana-schema/src/raw/composable`.

The big thing i had to fix in order to make this possible was the way that imports get generated in cuetsy. That's a nasty problem on its own, but will allow us to also solve the problem that core kinds needs to import the common schema area. So...worth it! grafana/cuetsy#85

i'll be doing another pass to whip this further into shape later.

Also, UGH, the jenny system really needs improvements in order to stop this silly duplication

### CODE REVIEW

**Review by dprokop — COMMENTED** [review]
Thanks @sdboyer for working on this <3

Can't provide much input on the codegen part. But thinking about the generated paths:

- Panel: `packages/grafana-schema/src/raw/composable/xychart/panelcfg/x/XYChartPanelCfg_types.gen.ts`
- Datasource: `packages/grafana-schema/src/raw/composable/tempo/dataquery/x/TempoDataQuery_types.gen.ts`


Is the `.../<panel-name>/panelcfg/....` and `.../<ds-name/dataquery/....` segment to provide consistency with already generated grok's go paths? Asking because when I think about it I don't see a different sub-namespaces(panelcfg, dataquery) coming. I think it would be more reasonable and semantic to reverse the path structure to be:

- Panel: `packages/grafana-schema/src/raw/composable/panel/xychart/x/XYChartPanelCfg_types.gen.ts`
- Datasource: `packages/grafana-schema/src/raw/composable/datasource/tempo/x/TempoDataQuery_types.gen.ts`

Mind the proposed structure: `.../<object-type>/<object-name>/x/XXX_types.gen.ts`. From discoverability POV I think it may be easier to think in object-types first when discovering the types, rather than exploring a long list of mixed composable types. 

WDYT?

**Comment by sdboyer** [comment]
That said, i am quite amenable to the object-types first approach. It would just mean using the compo kind name, not the plugin type, as the object type in the path element:

```
packages/grafana-schema/src/raw/composable/dataquery/tempo/v0/dataquery_types.gen.ts
packages/grafana-schema/src/raw/composable/dataquery/tempo/v1/dataquery_types.gen.ts
packages/grafana-schema/src/raw/composable/dataquery/tempo/v2/dataquery_types.gen.ts
*packages/grafana-schema/src/raw/composable/dataquery/tempo/v3/dataquery_types.gen.ts
*packages/grafana-schema/src/raw/composable/datasourcecfg/tempo/v0/datasourcecfg_types.gen.ts
```

**Review by torkelo — COMMENTED** [review]
How or the panel types exported? can't see anything exporting them, so not sure they will be included in the npm package / bundle?

**Comment by Clarity-89** [comment]
> currently they're not, that's the kind of thing i'm hoping for suggestions/discussions on. Main goal of this initial PR was to get just enough up in order to make topics like that concrete enough that we can tackle iteratively

We could namespace them when exporting, e.g. 
```
// index.gen.ts
export { PanelOptions as AlertGroupsPanelOptions } from "./raw/composable/alertgroups/panelcfg/x/AlertGroupsPanelCfg_types.gen.ts"
```
But maybe a better idea would be to have the same naming convention as for datasources, so instead of `PanelOptions`, they have the name of the panel in the definition - `AlertGroupsPanelOptions`, which would also make it more consitent with how we name the datasource queries.

**Comment by sdboyer** [comment]
@dprokop and i discussed the namespacing issue a bit when we first chatted about all this. We definitely have to do _something_ here; for an uncurated set of plugins, we simply can't aggregate all the exported symbols from all the plugins into the same namespace. Even if we're just thinking about the set of plugins owned by Grafana Labs, smashing all symbols into the same namespace creates a situation where e.g. [Cloudwatch's `MetricStat`](https://github.com/grafana/grafana/blob/0673c1a4eae8342fa8139671e786e8e8a59d9378/public/app/plugins/datasource/cloudwatch/dataquery.cue#L33) has to not collide with [Prometheus' `QueryEditorMode`](https://github.com/grafana/grafana/blob/0673c1a4eae8342fa8139671e786e8e8a59d9378/public/app/plugins/datasource/prometheus/dataquery.cue#L48). That's just a non-starter.

That really just leaves two options: either we concat the symbol names themselves as @Clarity-89 suggests, or we export them from separate modules within the same package.

While i can imagine it making sense for Scenes to adopt the former approach in the overall interface it presents (especially if y'all do curation of the included plugin set), at this layer, i'm inclined more towards separate modules. Concatenating names feels like:

* Repeatedly assaulting the developer's eyes with information in the body of the file they're writing context that was fully established up in the import statements
* If the importer wants a separate namespace within file scope, `import * from '@grafana/schema/composable/foo/panelcfg as foopanel' is trivial
* It produces type names that, for me, cross the line into [HighQualityEnterpriseNamingConventions](https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition) smelliness

(strong opinion, loosely held)

i realize that this generally runs afoul of TS conventions for wrapping everything in a package up at the top level. @dprokop looked at this a bit, excerpting from our DMs:

>> [@dprokop]: So I talked a bit with Jack about this last week. Gave him an overview what we are aiming for and this(namespacing within a single package) seems to be doable. There is tho a limitation on those types consumption if they are atomic modules within some higher level npm package (name it grafana/schema for now). And that’s related to typescript compiler config it that needs to have a specific (node16) module resolution set in order to be able to consume those namespaced modules.

> [@sdboyer] iiiiinteresting, i definitely had no idea about this one. am scanning over [the docs](https://www.typescriptlang.org/docs/handbook/module-resolution.html) now. It appears that setting this option isn’t leaky - so e.g., if scenes sets it to consume raw types, dependers on scenes won’t have to set it themselves

**Comment by dprokop** [comment]
> I think grouping the files in this way would suggest to any user that it's the plugin's version, not the kind's version, which is being represented in the path, leading to considerable confusion.

Yeah, I see the point. Given this, I think it's completely fine the way you proposed.

**Comment by dprokop** [comment]
> While i can imagine it making sense for Scenes to adopt the former approach in the overall interface it presents (especially if y'all do curation of the included plugin set), at this layer, i'm inclined more towards separate modules

I agree with @sdboyer about this one, but not might be not aware of all implications of TS namespacing and how much of an effort this will be.

Ideally I think we should have a standard convention for naming types like panel options or field config, I tried to describe this to @sdboyer in our private conversation:

___
I see one problem with the generated types as of now, that’s basically name clashes, i.e.
```import { PanelOptions as BarGaugePanelOptions } from 'app/plugins/panel/bargauge/panelcfg.gen';```
```import { PanelOptions as TablePanelOptions } from 'app/plugins/panel/table/models.gen';```

Mind the panel options types are named `PanelOptions`, which is inconsistent with i.e. field config types. I.e. `TableFieldOptions`, `GraphFieldConfig`, `PanelFieldConfig`(in Statetimeline panel)… I think there are probably more inconsistencies.
Ideally, consistency considered, I would expect the plugin to define panel options and field config types using the same name, would be perfect if at some point we had a validation for that!
I would consider such schema definition as std one for panel plugin:

```cue
composableKinds: PanelCfg: {
  maturity: "experimental"
  lineage: {
    seqs: [
      {
        schemas: [
          {
            Options: {
              …
            } @cuetsy(kind="interface")
            FieldConfig: {
              …
            } @cuetsy(kind="interface")
          },
        ]
      },
    ]
  }
}
```
Mind `Options` and `FieldConfig` in the above example. This of course forces us to start thinking about actual namespaced exports from schema: `import { Options, FieldConfig } from '@grafana
/schema/composable/timeseries'`.
___


Think it's worth involving @grafana/plugins-platform-frontend as I expect they may have plenty of input about structuring the export, and the expectations.

**Comment by sdboyer** [comment]
So there's two things here, i think. First is this:

> I would expect the plugin to define panel options and field config types using the same name, would be perfect if at some point we had a validation for that!

That's precisely what "[schema interfaces](https://docs.google.com/document/d/13Rv395_T8WTLBgdL-2rbXKu0fx_TW-Q9yz9x6oBjm6g/edit#heading=h.wf4l15jxalou)" do for us! They enforce/require certain fields to be present in all definitions of those composable kinds. The request here - removing the redundant `Panel` prefix - is straightforward enough: https://github.com/grafana/kindsys/issues/6. Requires updates to be made in a few repos, but it's an afternoon's work - though we should do _really_ soon, as it's very breaky.

The other is the names for generated TS symbols:

> This of course forces us to start thinking about actual namespaced exports from schema: `import { Options, FieldConfig } from '@grafana /schema/composable/timeseries'`.

i think the simplest option we have here is to expect the user to resolve this one by expecting that users[import into a variable](https://www.typescriptlang.org/docs/handbook/modules.html#import-the-entire-module-into-a-single-variable-and-use-it-to-access-the-module-exports):

```ts
import * as bargauge from '@grafana/schema/composable/panelcfg/bargauge/v1';
import * as table from '@grafana/schema/composable/panelcfg/table/v2';
```

AIUI this unfortunately goes a bit against TS standard practice, though.

i prefer this approach because it makes it trivial to know a priori what the name of the symbol will be. Contrast `FieldConfig` with `BarChartFieldConfig` vs `BarchartFieldConfig` - how would an automated tool know which naming capitalization pattern to pick? (i've started [a doc](https://docs.google.com/document/d/1xTRjsUpMrFTJJx67LjxRHz-l7JnIdmFDLxojClR11yE/edit#heading=h.5sybau7waq2q) to deal with questions like this because it's a pervasive and nasty problem, but there's nothing really in there yet)

**Comment by mckn** [comment]
Great initiative with this draft PR @sdboyer! A couple of things that I was thinking about when reading the comments and going through this code.

I saw that we talked about using some kind of namespacing to scope the types and prevent collisions on the exposed scope of the package. I hope you wasn't referring to TS namespace because I don't think it would be wise to start use that feature of typescript. See [here](https://devblogs.microsoft.com/typescript/typescripts-migration-to-modules/) for more information.

Placing all the generated types in the `@grafana/schemas` package makes it harder to extract our core plugins away from core. We have a dna doc and have started to look into different ways to approach this work. With that said I'm not sure when we actually will start this work but we are starting to look into different ways to do it.

I'm a bit confused when it comes to the versioning part of the generated paths. I am probably missing something obvious but here we go...why do we need to have multiple versions of the same data query type available in the package?

```
packages/grafana-schema/src/raw/composable/dataquery/tempo/v0/dataquery_types.gen.ts
packages/grafana-schema/src/raw/composable/dataquery/tempo/v1/dataquery_types.gen.ts
``` 

Instead of adding the generated types to the `@grafana/schemas` package. Could we instead have a private repository where the type definitions live which would published to NPM as separate packages? That would solve the issue with collisions in the export scope. Also, it would enable us to include external plugins in the future as well. It would also enable us to correlate the version of the kind with the version of the published package since each plugin would be published as a separate package.

I'm thinking that it could work similar to how the https://github.com/DefinitelyTyped/DefinitelyTyped does it?

**Comment by sdboyer** [comment]
thanks for the feedback, @mckn! going through...

> I hope you wasn't referring to TS namespace because I don't think it would be wise to start use that feature of typescript. See [here](https://devblogs.microsoft.com/typescript/typescripts-migration-to-modules/) for more information.

Definitely referring to the general concept of a namespace, not any particular TS construct. Though if i'm understanding that blog post correctly, it suggests that what we're considering doing here - creating a set of modules that _amount to_ a logical namespace - aligns with a golden path that the TS folks are trying to create in the compiler itself.

> Placing all the generated types in the @grafana/schemas package makes it harder to extract our core plugins away from core.

This is something @dprokop and i talked about quite a bit.

It's not at all intrinsically obvious from the PR, but this actually _doesn't_ make it harder. All of this is created by jennies, and what we're doing here is experimenting with things like the output path structure with the understanding that it will be trivial to pick up the jenny and move it into grok, which is where this ultimately belongs. So, to the other points:

> I'm a bit confused when it comes to the versioning part of the generated paths. I am probably missing something obvious but here we go...why do we need to have multiple versions of the same data query type available in the package?

The primary audience for all generated code in grok is users _outside_ of Grafana. This is obviously the case for target tools like jsonnet/grafonnet or especially terraform, where it's really only going to be useful for more traditional infrastructure as-code. But the key insight @dprokop had is that part of what scenes is aiming to do - a declarative interface to building Grafana - _is not fundamentally different to what we're doing in grafonnet_: providing a convenient, language-native target for declaratively configuring Grafana.

Given that, we're thinking of the generated output in this PR as the lowest level, raw generated TS output, which will eventually move into grok, and then be composed together by scenes, in a similar manner to how grafonnet composes together the lower-level jsonnet output [in grok today](https://github.com/grafana/grok/tree/main/jsonnet/v9.4.0).

SO, the reason we have the internally versioned paths here is the same reason we have them elsewhere in grok - because it provides stable, non-breaking import paths for consuming these types, including importing multiple versions of them simultaneously when stitching together a library such as scenes.

So, to this:

> Instead of adding the generated types to the @grafana/schemas package. Could we instead have a private repository where the types live which would be published to NPM as separate packages?

Our plan going forward is to have a schema registry is to have a separate repository which contains (copies of) the original CUE definitions of kinds  - [proposal 2, here](https://docs.google.com/document/d/1kQ3mFdtFeF569Icbg2c4yTmCRvnKhW6a8WRCf6EQmkI/edit#heading=h.f3nnjfiyz6v0). This will be essential for pulling plugins out of core grafana - without such a registry, they'd no longer show up in grok, as grok currently can only pull kinds defined in grafana/grafana.

The registry for CUE definitions is decided at this point, but we do have flexibility around how we generate the types themselves. We certainly could choose to make another jenny that's designed to support publishing distinct npm packages. Probably deserves exploration in its own design doc. This approach, though, seems like it's the best substrate for building something like scenes.

**Comment by torkelo** [comment]
Moving core plugins out of Grafana? That sounds messy

**Comment by torkelo** [comment]
But reading the design doc it makes some sense, but worried about timeline. For the core visualization/ can we find a simple temporary solution?

**Comment by sdboyer** [comment]
Yeah the main motivation for the schema registry is providing assurances in CI, not the plugins-outside-core case. (Yes, it's difficult - that's why it's a north star.) Rather, it's just one that's enabled by it.

This PR is the temporary solution. All the above discussion is just explaining how this temporary solution _also_ provides a smooth transition towards the long-term goal.

**Comment by mckn** [comment]
> Moving core plugins out of Grafana? That sounds messy

Yes - I totally agree! We are currently trying to figure out where this requirement comes from and why we want to do it. FYI: It isn't our squad pushing for this.

**Comment by torkelo** [comment]
any chance to get this into G10 ? all the users of scenes are asking for typed visualizations

File: lerna.json
Comment (tolzhabayev) [inline lerna.json:4]: seems unrelated? auto changed?

**Comment by sdboyer** [comment]
i think this is good to go, unless we want to wait until after we've landed grafana/kindsys#7 and the related grafana/grafana#65297 - @dprokop, are you actively blocked on this in a way where not having to wait a few days will make a difference?

**Comment by dprokop** [comment]
> @dprokop, are you actively blocked on this in a way where not having to wait a few days will make a difference?

Nope, we can wait a few days, don't worry. Not actively working on API atm. Will circle back to this PR ltr this week.

**Comment by spinillos** [comment]
@dprokop could you re-check if everything is ok 👀? I just re-generated everything with the new changes.

**Comment by dprokop** [comment]
@spinillos I took a look at the generated ts files and stuff looks good!

**Review by dprokop — APPROVED** [review]

**Pull request merged**
